feat!: Merge with degree check

federicobarbacovi opened the pull request:
We modify the Merge protocol so that it enforces that the subtable polynomial `t_j` has degree smaller than `subtable_size` as read from the proof.

Closes https://github.com/AztecProtocol/barretenberg/issues/1442

**Details**

As per the linked issue, we want to support ecc operations in app circuits. To ensure that app ecc ops do not modify ecc ops that happened before them, we need to ensure that the subtable length `t.size()` is smaller than the constant `l` by which we right shift `T_prev`. This is to ensure that `t + X^{l-1} T_prev` is indeed the polynomial corresponding to the column `t || T_prev`.

We enforce this degree check in the merge protocol by requiring the prover to commit to `g(X) := X^{l-1} t(1/X)` and provide openings `c`, `d` of `t`, `g` at challenges `1/kappa`, `kappa`, respectively, for which we check `c * kappa^{l-1} = d`.

To save on the number of MSMs performed, we use Shplonk to verify the following claims:
- `t(X)` opens to `c` at `1/kappa`
- `p(X) = t(X) + X^{l-1} T_prev(X) - T(X)` opens to `0` at `kappa`
- `g(X) := X^{l-1} t(1/X)` opens to `d` at `kappa`

Commit:
[empty] Start merge-train. Choo choo.

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

iakovenkos committed:
chore: stdlib bool internal audit  (#15070)

* Our basic binary ops produce a new gate iff both operands are
witnesses. Most of the implementations contained deeply branched logic
that is now replaced with simple `int` arithmetic that looks more robust
and is much easier to read.

* Added documentation in various places. The constraints are made more
explicit
* Removed redundant assignments
* Re-worked the test suite. Now all basic ops and functions are tested
for all combinations of flags (`is_constant`, `witness_bool`,
`witness_inverted`). Added missing tests, checked various edge cases.
* Removed a couple of tests that seemed less expressive
* Removed `must_imply()` method for a vector of implications, as it is
not used anymore
* Fixed a `normalize()` regression - previously it wouldn't check if
(`witness_inverted == false`) and would normalize a witness `bool_t`
even if the condition is satisfied. It is the reason for the VK changes

federicobarbacovi committed:
Comments

federicobarbacovi committed:
feat: improve Shplonk api (#15422)

We enhance the Shplonk verifier api so that it can efficiently handle
openings of polynomials that are linearly dependent

---------

Co-authored-by: AztecBot <tech@aztecprotocol.com>
Co-authored-by: ludamad <adam.domurad@gmail.com>
Co-authored-by: Suyash Bagad <suyash@aztecprotocol.com>
Co-authored-by: sergei iakovenko <105737703+iakovenkos@users.noreply.github.com>
Co-authored-by: Jonathan Hao <jonathan@aztec-labs.com>
Co-authored-by: ledwards2225 <98505400+ledwards2225@users.noreply.github.com>

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
Restructure merge prover, verifier and recursive verifier

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

ludamad committed:
fix(merge-train): don't queue merge if merge-train failed queue, pass on rebase logic (#15508)

iakovenkos committed:
chore: nuke bit array (#15522)

The `bit_array` stdlib primitive is not used in the production code -->
deleted

Had to change a couple of headers that included `bit_array` to
access`plookup` or `uint`.

Fixed some misleading comments in a couple of fuzzers

iakovenkos committed:
chore: remove template parameters (#15530)

`ECCVM-` and `Translator-` `RecursiveFlavors` are only used with
`UltraCircuitBuilder`, which allows us to decouple them and the
corresponding `RecursiveVerifiers` from `Flavor` template parameter.

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

ledwards2225 committed:
chore: no PK (#15386)

Removes the `ProvingKey` altogether from the Ultra/Mega proving systems.

TLDR: "proving key" used to mean "the precomputable data used by the
prover." The Honk PK hasn't been that for a long time since it stores
the `ProverPolynomials` which contains _all_ polynomials used by the
prover, including witnesses. This meant there was no clear logic behind
what was stored in the `DeciderProvingKey` vs the `ProvingKey`. Since we
don't have much use for the traditional notion of a proving key in Honk
(unlike in Plonk), I've opted to remove it entirely and place its data
directly in the `DeciderProvingKey`.

Most lines in this PR relate to one of these changes:
- There is a single circuit-size-like parameter in the DPK, accessed via
`dyadic_size()`
- A honk VK is now constructed as `vk(pk->get_precomputed())` (instead
of `vk(pk->proving_key)`)
- Lines containing `proving_key->proving_key.foo` have been reduced to
`proving_key->foo`
- "metadata" (circuit size and public inputs metadata) live in a
`MetaData` struct. (Note: I'm open to a better name..)

---------

Co-authored-by: AztecBot <tech@aztecprotocol.com>
Co-authored-by: ludamad <adam.domurad@gmail.com>
Co-authored-by: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Co-authored-by: Suyash Bagad <suyash@aztecprotocol.com>
Co-authored-by: sergei iakovenko <105737703+iakovenkos@users.noreply.github.com>

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
Improve code

federicobarbacovi committed:
Precompute inverse vanishing evals

federicobarbacovi committed:
Extend tests; fix verifiers' algorithms

federicobarbacovi committed:
Fixes; extend tests

federicobarbacovi committed:
Cleanup; update constants

federicobarbacovi committed:
Comments

ledwards2225 committed:
chore!: Correct public inputs propagation in the tube (#15547)

The tube propagates the public inputs of the hiding circuit (received in
the CIVC proof) via its own public inputs. Previously, those public
inputs were being extracted from the native proof meaning they were
disconnected from those used by the CIVC recursive verifier. A previous
PR made the CIVC rec verifier take a stdlib proof so it is now easy to
directly set the pub inputs of the stdlib proof public, making the
connection proper.

Partially (entirely?) addresses
https://github.com/AztecProtocol/barretenberg/issues/1048

federicobarbacovi committed:
Fix boomerang variables

notnotraju committed:
chore: use `batch_invert` in native IPA verifier (#15557)

Addresses
[1140](https://github.com/AztecProtocol/barretenberg/issues/1140),
native IPA verifier computes inverses of the challenges via a
`batch_invert()` call.

---------

Co-authored-by: notnotraju <raju@aztec-labs.com>

federicobarbacovi committed:
Refactoring

suyash67 committed:
chore: Move `stdlib::uint_plookup` to `stdlib::uint` (#15460)

Moves the `uint_plookup` class into the `uint` class in stdlib. This is
done because the plookup-based uint version is now the default (since we
no more the old `standard` or `turbo` arithmetisation).

As a result, we remove any conditionals using `HasPlookup` from uint
related code.

federicobarbacovi committed:
Clean up

johnathan79717 committed:
chore: use const ref commitment keys (#15584)

I had thought that const ref can't get a default parameter but it can!

ludamad committed:
Merge remote-tracking branch 'origin/next' into merge-train/barretenberg

ludamad committed:
Merge commit '42b65c12b46f46aac143fb8792440e1ea3720218' into merge-train/barretenberg

iakovenkos committed:
fix: hiding circuit vk computed only once (#15589)

removed redundant hiding circuit vk computation

Commit:
Merge branch 'next' into merge-train/barretenberg

lucasxia01 committed:
feat: transcript can hash objects independently (#15510)

Adds new transcript functionality that allows us to hash the vk (and
soon the accumulator) using an independent buffer in the transcript to
allow for the OriginTags to still be correctly assigned and also having
the hash of the object be independent of the transcript state. The hash
of the object is added to the actual hash buffer that is used to
generate challenges in the protocol.

---------

Co-authored-by: AztecBot <tech@aztecprotocol.com>

johnathan79717 committed:
chore: readme for benchmarking remotely (#15512)

Writing stuff down since Raju had the same problem trying to benchmark
remotely as I did. It may benefit someone else or myself in the future.
I left out the instance URL as I'm not sure if it should be shared
publicly.

ledwards2225 committed:
chore: fix avm test (#15592)

Please read [contributing guidelines](CONTRIBUTING.md) and remove this
line.

For audit-related pull requests, please use the [audit PR
template](?expand=1&template=audit.md).

notnotraju committed:
chore: hash more stuff for IPA. (#15519)

Adds a small `add_to_hash_buffer` method for `MockTranscript` and then
has the IPA prover add the commitment, challenge, and evaluation to the
hash buffer.

Addresses
[1150](https://github.com/AztecProtocol/barretenberg/issues/1150) and at
least part of
[1408](https://github.com/AztecProtocol/barretenberg/issues/1408). See
the commentary on the former for why this solution is the simplest/most
ergonomic for soundness of IPA.

---------

Co-authored-by: notnotraju <raju@aztec-labs.com>
Co-authored-by: ludamad <adam.domurad@gmail.com>

lucasxia01 committed:
chore: fix avm build in merge-train/bb (#15594)

Accidentally broke build by redefining to_field_elements in avm VK

Commit:
Merge branch 'next' into merge-train/barretenberg

ledwards2225 committed:
feat!: structured public inputs via kernel io (#15383)

Introduces the `KernelIO` class/mechanism to begin unifying the handling
of certain fixed internal public inputs.
- For now it's utilized only for the Databus return data commitments
- Removes all use of `app/kernel_return_data_commitment_pub_input_key`
in DataBusPropagationData (but they are not yet removed from the VKs)
- Confirmed that changes in VKs of certain kernels is due only to the
above change (public component keys are still present but unset so take
default values)

Closes https://github.com/AztecProtocol/barretenberg/issues/1371

---------

Co-authored-by: AztecBot <tech@aztecprotocol.com>
Co-authored-by: ludamad <adam.domurad@gmail.com>
Co-authored-by: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Co-authored-by: Suyash Bagad <suyash@aztecprotocol.com>
Co-authored-by: sergei iakovenko <105737703+iakovenkos@users.noreply.github.com>
Co-authored-by: Jonathan Hao <jonathan@aztec-labs.com>
Co-authored-by: Khashayar Barooti <khashayar@aztecprotocol.com>

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

ledwards2225 committed:
fix!: aggregate correct nested pairing points in the hiding circuit (#15598)

Prior to this fix, the nested pairing points from the tail kernel were
not being propoerly extracted in the hiding circuit. Instead of
extracting them from the public inputs of the tail decider VK, we were
extracting points from the accumulator VK (whose public inputs are not
meaningful).

Note: this changes the VK only for the hiding circuit which is not
captured by our VK consistency checks

iakovenkos committed:
fix: bb merge-train conflicts (#15617)

conflict resolution cause my PR #15427 stopped the train

---------

Co-authored-by: Jean M <132435771+jeanmon@users.noreply.github.com>
Co-authored-by: Alex Gherghisan <alexghr@users.noreply.github.com>
Co-authored-by: Santiago Palladino <spalladino@users.noreply.github.com>
Co-authored-by: Santiago Palladino <santiago@aztec-labs.com>

federicobarbacovi committed:
chore: Refactor shplonk verifier api (#15618)

Refactor the Shplonk verifier api:
* Remove `OpeningVector` (which was used only in Sphlonk) class in
favour of `ShplonkVerifier::UpdateData`
* Refactor methods using `OpeningVector` accordingly
* Refactor `ShplonkVerifier::update` so that it uses a single opening
pair `(x, v)` instead of a challenge `x` and a series of evaluations
`(v_1, .., v_k)` (as it should be, given that the assertion of the
verifier will only concern the `v` and not the single evaluations `v_1,
.., v_k`).

federicobarbacovi committed:
Merge branch 'merge-train/barretenberg' into fb/merge_with_degree_check

lucasxia01 committed:
Merge remote-tracking branch 'origin/next' into merge-train/barretenberg

federicobarbacovi committed:
Refactor

lucasxia01 committed:
Merge remote-tracking branch 'origin/next' into merge-train/barretenberg

federicobarbacovi committed:
Clean up

federicobarbacovi committed:
Update tests

ledwards2225 committed:
chore!: databus consistency checks in the hiding circuit (#15599)

Uses the new KernelIO mechanism to add databus consistency checks to the
hiding circuit. Shoutout to @kashbrti for figuring out exactly what
needed to be done here.

Note: we probably can actually get away with only performing one
consistency check in the hiding circuit but since its basically free to
do both I'll leave it that way for now. This is consistent with all
other kernels

Closes: https://github.com/AztecProtocol/barretenberg/issues/1397

lucasxia01 committed:
feat!: VK hash consistency check (#15591)

Checks that the vk hash being passed in matches the hash of the
verification key that's generated from the transcript in the Oink
recursive verifier. Also updates the VK hash to be 254 bits instead of
128 bits.

Closes https://github.com/AztecProtocol/barretenberg/issues/1447.

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

ludamad committed:
[empty] Start train.

federicobarbacovi committed:
Merge branch 'merge-train/barretenberg' into fb/merge_with_degree_check

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/merge_with_degree_check

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

maramihali commented on barretenberg/cpp/src/barretenberg/ultra_honk/merge_verifier.cpp:
```diff
@@ -13,86 +13,180 @@ namespace bb {
-        T_prev_evals[idx] = transcript->template receive_from_prover<FF>("T_prev_eval_" + std::to_string(idx));
+        // Evaluation is hard-coded to zero
+        ShplonkVerifier::LinearCombinationOfClaims claim{
+            { idx + t_IDX * NUM_WIRES, idx + T_PREV_IDX * NUM_WIRES, idx + T_IDX * NUM_WIRES },
```
I am missing a bit of context as to why these indices are needed but I wonder whether we cannot find a cleaner way to do this as part of Shplonk rather than introducing this inter-dependence between the two protocols. Not something to be handled in this PR but we should revisit (and in the context of this PR, some more documentation would be great regarding to what the claims sent to Shplonk mean). Happy to discuss more!

federicobarbacovi replied:
The indices are needed to identify which commitments the scalar coefficients corresponds to. There is a method in the Shplonk verifier that doesn't require exposing these indices, but that would require precomputing the commitment to `p_j(X) = l_j(X) + X^k r_j(X)`, so in this case it would be costly. Happy to discuss more after this PR!

maramihali commented on barretenberg/cpp/src/barretenberg/ultra_honk/merge_verifier.cpp:
```diff
@@ -13,86 +13,180 @@ namespace bb {
+     *      [t_1], .., [t_4], [T_{1,prev}], .., [T_{4, prev}], [T_1], .., [T_4], [g_1], .., [g_4]
+     * }
+     *
+     * And then we open the claims by linearly combining the commitments:
```
This matrix is a bit hard to follow when you're having a first look at his, maybe you can present this piece of info in a different way :)

Commit:
feat: merge protocol correctly processes both appending and prepending of incoming tables (#15622)

Update the EccOpQueue and Merge protocol functionality to allow both
prepending and appending new subtables.

The motivation of this is to ensure that the last merge proof, sent to
the rollup, cannot enable the verifier to find out the number of ecc ops
in the transaction.

maramihali reviewed (changes requested):
Thanks for your work, really appreciate the thorough testing in the stdlib universe and I think it would be great to have  similar tests in the native universe as well! I think this PR is quite large and it would have benefited from being separated into several PRs also when it comes to reviewing as the diffs basically encompass entire files (something to think of doing in future pieces of work). I am not a big fan of the indexing using constants into an array of `table_polynomials` to get T, T_prev, t_curr,.. so I think the design of that is worth re-visiting. Also I think the inter-dependence between MergeVerifier and ShplonkVerifier can be reduced, I imagine that part can be refactored so that MergeVerifier prepares the opening claims to be sent to Shplonk and then Shplonk handles intermediate computation rather than having some done in MergeVerifier and some in ShplonkVerifier which really affects the readability of the protocol as a whole. I also requested more documentation when it comes to what exactly the opening claims sent to Shplonk represent (the `indices` part in LinearCombinationOfClaims` is also something that I think would benefit from a revisiting).

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
Address minor comments

federicobarbacovi commented on barretenberg/cpp/src/barretenberg/commitment_schemes/shplonk/shplonk.hpp:
```diff
@@ -390,6 +390,13 @@ template <typename Curve> class ShplonkVerifier_ {
+     *
+     * @return Fr
+     */
+    Fr get_z_challenge() { return z_challenge; }
```
Removed the getter as we agreed it was not worth the increased messiness in the Merge implementation. `z` needs to be a member variable for the Shplonk verifier to work in a streaming fashion (with the internal state updated via successive calls of the `update` method)

federicobarbacovi committed:
Removed shplonk getter and simplified rec merge verifier code

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
Prover stores tables

federicobarbacovi committed:
Revert prover to monolithic structure

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
Revert native verifier to monolithic structure

federicobarbacovi committed:
Return rec merge verifier to monolithic structure

federicobarbacovi committed:
Fixes

ledwards2225 committed:
chore!: use kernel/app io mechanism in CIVC (#15603)

Completes the integration of `KernelIO`/`AppIO` into CIVC so that
`PublicComponentKey`s are no longer required.
- Integrates pairing points into KernelIO
- Introduces `AppIO` (just pairing points)
- Does not yet remove any `PublicComponentKey`s from the VKs (to be
handled in a follow on)

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
Comments

federicobarbacovi commented on barretenberg/cpp/src/barretenberg/ultra_honk/merge_prover.hpp:
```diff
@@ -42,9 +53,29 @@ class MergeProver {
+     */
+    std::array<std::array<Polynomial, NUM_WIRES>, 4> preamble_round();
+
+    /**
```
Moved to the cpp file

federicobarbacovi committed:
Remove unused code

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
Improve comments for merge prover

federicobarbacovi committed:
Improve verifier's comments

federicobarbacovi committed:
Comments

federicobarbacovi commented on barretenberg/cpp/src/barretenberg/stdlib/merge_verifier/merge_recursive_verifier.cpp:
```diff
@@ -15,21 +15,141 @@ MergeRecursiveVerifier_<CircuitBuilder>::MergeRecursiveVerifier_(CircuitBuilder*
+    // Update state of the verifier
+    static constexpr size_t NUM_KAPPA_INV_CLAIMS = 4;
+    for (size_t idx = 0; idx < NUM_KAPPA_INV_CLAIMS; ++idx) {
+        verifier.update(opening_claims[idx], kappa_inv_vanishing_eval);
```
Restructured as discussed. The savings from extracting `z_challenge` were not worth it, so we now pass the claims directly to Shplonk, which handles the computation of `1 / (z_challenge - kappa)` and `1 / (z_challange - kappa_inv)`

lucasxia01 committed:
feat: fiat-shamir VK in ECCVM and Translator (#15640)

FS the VKs in the ECCVM and Translator.

Closes https://github.com/AztecProtocol/barretenberg/issues/1466.

ludamad committed:
chore(bb): instrument tracy fully (#14820)

Closing https://github.com/AztecProtocol/aztec-packages/pull/14754 as I
realized that I didnt properly pluck it out of the memory branch I had.
This includes everything needed to make it work (memusage script itself
already reviewed)

Including everything needed to have (almost? maybe 30mb less than
`heaptrack` reports now) full memory accounting from tracy, finding an
unaccounted for 150mb+ thru aligned operator new (though none of the
operator new overloading seemed to work prior)

![Screenshot 2025-06-05 at 12 50
37 PM](https://github.com/user-attachments/assets/8b85badb-ef04-4a49-b770-62c9432d89c6)

ludamad committed:
chore: remove large vk constructors meant for msgpack (#15654)

This was just an additional safety check. Where these constructors are
not otherwise used it can be tedious to have, its more meant to
encourage 'no explicit constructor' structs. This adds a bypass. Closes
https://github.com/AztecProtocol/barretenberg/issues/964

Commit:
Merge branch 'next' into merge-train/barretenberg

lucasxia01 committed:
chore: ECCVM and Translator VK/Proof tests (#15658)

Implements to_field_elements() for ECCVM and Translator VKs and adds
more tests.

ledwards2225 committed:
fix: correct pub inputs propagation in hiding kernel (#15657)

Extract the tail kernel public inputs from the stdlib proof instead of
the native proof to correctly connect the underlying witnesses.

Commit:
Merge branch 'next' into merge-train/barretenberg

ludamad committed:
chore(msgpack): vk boilerplate reduction (#15659)

Fixes the rest of barretenberg/964

Closes https://github.com/AztecProtocol/barretenberg/issues/964

ludamad committed:
feat(bbapi): rename bbrpc=>bbapi, move CIVC CLI to use bbapi (#15502)

Start of UH bbapi implementation, make the client ivc test use bbapi
Note that the file names are clashing here, i will move the original
api_ultra_honk etc to cli_ultra_honk files soon

ledwards2225 committed:
chore: remove DatabusPropagationData altogether (#15611)

Satisfying!
- removes `DatabusPropagationData` altogether (made possible by
`KernelIO` mechanism)
- `is_kernel` is now a member of the builder only (not in VK)
- `is_kernel` is stored in the verification queue entries (set based on
`circuit.is_kernel` in `accumulate()`)

closes: https://github.com/AztecProtocol/barretenberg/issues/1454
closes: https://github.com/AztecProtocol/barretenberg/issues/1371

Commit:
Merge branch 'next' into merge-train/barretenberg

ludamad committed:
Revert "feat(bbapi): rename bbrpc=>bbapi, move CIVC CLI to use bbapi" (#15667)

Reverts AztecProtocol/aztec-packages#15502

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/merge_with_degree_check

federicobarbacovi committed:
Comment

federicobarbacovi committed:
Minor changes to restart working

federicobarbacovi committed:
feat: KZG handles empty polynomial (#15652)

* KZG prover handles the empty polynomial by interpreting it as the zero polynomial.
* Pairings handle points at infinity in G1 and G2

Closes https://github.com/AztecProtocol/barretenberg/issues/79

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
WIP

federicobarbacovi committed:
Update merge provers/verifiers

federicobarbacovi committed:
Update tests

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/merge_with_degree_check

federicobarbacovi committed:
Fix

federicobarbacovi committed:
Fix pow; remove append only merge test

federicobarbacovi committed:
Clean up

Commit:
[empty] Start merge-train. Choo choo.

maramihali reviewed (commented):
thanks for the modifications, another bulk of comments but mostly nits! :)

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/merge_with_degree_check

federicobarbacovi committed:
Revert changes to mega flavor

federicobarbacovi commented on barretenberg/cpp/src/barretenberg/stdlib/merge_verifier/merge_recursive_verifier.cpp:
```diff
@@ -18,98 +18,157 @@ MergeRecursiveVerifier_<CircuitBuilder>::MergeRecursiveVerifier_(CircuitBuilder*
-        std::string suffix = std::to_string(idx);
-        T_prev_commitments[idx] = transcript->template receive_from_prover<Commitment>("T_PREV_" + suffix);
-        T_commitments[idx] = transcript->template receive_from_prover<Commitment>("T_CURRENT_" + suffix);
+        if (settings == MergeSettings::PREPEND) {
```
They add the commitments to `T_prev` in different order (if `PREPEND`, first `t_commitments` then `T_prev`, if `APPEND`, then the reverse)

federicobarbacovi replied:
No, the if statement will stay, but instead of receiving the data from the prover we will have it as an input to the proof verification. Happy to discuss more if you want :) (I replaced the if with ternary statements, they look better)

federicobarbacovi replied:
No, the if statement will stay, but instead of receiving the data from the prover we will have it as an input to the proof verification. Happy to discuss more if you want :) (I replaced the if with ternary statements, they look better)

federicobarbacovi commented on barretenberg/cpp/src/barretenberg/ultra_honk/merge_verifier.cpp:
```diff
@@ -16,94 +16,158 @@ MergeVerifier::MergeVerifier(const std::shared_ptr<Transcript>& transcript, Merg
+        commitment_idx += 1;
+
+        // Degree identity
+        degree_check_verified &= (left_table_eval_kappa_inv * pow_kappa_minus_one == left_table_reversed_eval);
```
Add an issue https://github.com/AztecProtocol/barretenberg/issues/1475

federicobarbacovi committed:
Address comments

federicobarbacovi commented on barretenberg/cpp/src/barretenberg/stdlib/merge_verifier/merge_recursive_verifier.cpp:
```diff
@@ -18,98 +18,157 @@ MergeRecursiveVerifier_<CircuitBuilder>::MergeRecursiveVerifier_(CircuitBuilder*
+                        /*index of [m_j]*/ commitment_idx + 2 },
+                      { FF(1), pow_kappa, FF(-1) },
+                      { kappa, FF(0) } };
+        opening_claims.emplace_back(claim);
```
Thanks!

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/merge_with_degree_check

federicobarbacovi committed:
Update mock verifier inputs

maramihali reviewed (approved):
thanks for bearing through my endless review :D

federicobarbacovi commented on barretenberg/cpp/src/barretenberg/ultra_honk/merge_verifier.cpp:
```diff
@@ -36,44 +61,12 @@ MergeVerifier::MergeVerifier(const std::shared_ptr<Transcript>& transcript, Merg
     transcript->load_proof(proof);
 
     const uint32_t shift_size = transcript->template receive_from_prover<uint32_t>("shift_size");
+    ASSERT(shift_size > 0, "Shift size should always be bigger than 0");
```
Thanks!

federicobarbacovi committed:
Style
